astropy/modeling/physical_models.py

```diff
@@ -371,3 +372,334 @@ def input_units(self):
+        self.radius_s = self.radius_s.to(u.kpc)
+
+        # Calculate scale density
+        self.density_s = in_mass / (4.0 * np.pi * self.radius_s ** 3 * self.A_NFW(concentration))
+
+        super().__init__(mass=in_mass, concentration=concentration, redshift=redshift, **kwargs)
+
+    def evaluate(self, r, mass, concentration, redshift):
+        """
+        One dimensional NFW profile function
+
+        Parameters
+        ----------
+        r : float or :class:`~astropy.units.Quantity`
+            Radial position of density to be calculated for the NFW profile.
+        mass : float or :class:`~astropy.units.Quantity`
+            Mass of NFW peak within specified overdensity radius.
+        concentration : float
+            Concentration of the NFW profile.
+        redshift : float
+            Redshift of the NFW profile.
+
+        Returns
+        -------
+        density : float or :class:`~astropy.units.Quantity`
+            NFW profile mass density at location ``r``. The density units are:
+            [``mass`` / ``r`` ^3]
+
+        Notes
+        -----
+        .. warning::
+
+            Output values might contain ``nan`` and ``inf``.
+        """
+        # Enforce default units
+        if not isinstance(mass, u.Quantity):
+            in_mass = u.Quantity(mass, u.M_sun)
+        else:
+            in_mass = mass
+
+        # Create radial version of input with dimension
+        if hasattr(r, "unit"):
+            in_r = r
+            self.radius_s = self.radius_s.to(r.unit)
+            self.density_s = self.density_s.to(in_mass.unit / r.unit**3)
+        else:
+            in_r = u.Quantity(r, u.kpc)
+
+        # Update NFW Scale radius
+        # Delta Mass is related to delta radius by
+        # M_{200}=\frac{4}{3}\pi r_{200}^3 200 \rho_{c}
+        # And delta radius is related to the NFW scale radius by
+        # c = R / r_{\\rm s}
+        self.radius_s = (((3.0 * in_mass) / (4.0 * np.pi * self.density_delta))
+                         ** (1.0 / 3.0)) / self.concentration
+        self.radius_s = self.radius_s.to(in_r.unit)
+
+        # Update NFW Scale density
+        self.density_s = in_mass / (4.0 * np.pi * self.r_s ** 3 * self.A_NFW(self.concentration))
+
+        # Define reduced radius (r / r_{\\rm s})
+        radius_reduced = in_r / self.r_s.to(in_r.unit)
+
+        # Density distribution
+        # \rho (r)=\frac{\rho_0}{\frac{r}{R_s}\left(1~+~\frac{r}{R_s}\right)^2}
+        density = self.rho_s / (radius_reduced * (u.Quantity(1.0) + radius_reduced) ** 2)
+
+        if hasattr(mass, "unit"):
+            return density
+        else:
+            return density.value
+
+    @staticmethod
+    def A_NFW(y):
+        r"""
+        Commonly used expression in several calculations.
+
+        Notes
+        -----
+
+        Model formula:
+
+        .. math:: A_{NFW} = [\ln(1+y) - \frac{y}{1+y}]
+        """
+        return np.log(1.0 + y) - (y / (1.0 + y))
+
+    @property
+    def rho_s(self):
```

Comment:
Is this name common among users of this model? To me `scale_density` or `rho_scale` sounds clearer. I'll differ to @eteq on this one.

Reply:
This is the name commonly used in the literature, but I am happy to change it if @eteq so desires.

Reply:
I think for this one I agree with @nden and I like `rho_scale`, because this is not quite as universally used as `r_s`.  But I think it's probably good to add some text like ``often written in the literature as :math:`\rho_s` `` or something like that?

Reply:
Implemented as per Eric's comment.